### PR TITLE
36 pagination courses

### PR DESCRIPTION
**Pull request by MikeVarchola**
@jonbiemond 
Limit courses displayed to 25 per page - DONE

I added URL params for filtering and pagination via GET request
Added tests for the pagination

### CODE REVIEW

File: bcitflex/templates/courses/index.html
Comment (jonbiemond) [inline bcitflex/templates/courses/index.html:9]: Could you please share why the switch to GET? Forgive my ignorance here, so feel free to link me to a resource.

File: bcitflex/templates/courses/index.html
Comment (jonbiemond) [inline bcitflex/templates/courses/index.html:24]: This is an awesome improvement!

File: bcitflex/course.py
Comment (jonbiemond) [inline bcitflex/course.py:33]: Any particular reason to switch from `select` to `query`? As of SQLAlchemy 2.0 the [Query API has been marked as legacy](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#legacy-query-api).

File: bcitflex/app_functions/course_query.py
Comment (jonbiemond) [inline bcitflex/app_functions/course_query.py:1]: Unfortunately I don't think this change will work with filters on attributes of a model that are not columns in the database.

**Review by jonbiemond — CHANGES_REQUESTED** [review]
This looks awesome @MikeVarchola! Could you please look into why the tests might be failing? I've also added some comments asking for clarification on a couple of your changes.

File: bcitflex/app_functions/course_query.py
Comment (jonbiemond) [reply bcitflex/app_functions/course_query.py:1]: Yes, I like the idea of a view, I think it would be neater. But I don't like how it abstracts some of the logic into SQL and out of Python. I'd love to discuss this more. Is this change needed to implement the Flask-SQLAlchemy pagination?

Comment (MikeVarchola) [reply bcitflex/app_functions/course_query.py:1]: Well when you look at your code, you select all the records from Courses table, and iterate them and filter them. I think that it's better performance wise to use filters, sorting and pagination in DB. But if you insist I can take a look if it's possible to make pagination on Python level with your code.

Comment (MikeVarchola) [reply bcitflex/app_functions/course_query.py:1]: @jonbiemond Flask-SQLAlchemy pagination is static method mainly used by Flask-SQLAlchemy query, which I used, there is no other option to imlement it.

File: bcitflex/course.py
Comment (MikeVarchola) [reply bcitflex/course.py:33]: @jonbiemond  `query` implements static method .paginate() from https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/pagination/

there is no other implementation of SQL-Alchemy Pagination

File: bcitflex/templates/courses/index.html
Comment (MikeVarchola) [reply bcitflex/templates/courses/index.html:9]: @jonbiemond  GET is best practice for getting list of objects out of DB, using query and queryParams. Also for implementing requested pagination.

File: bcitflex/app_functions/course_query.py
Comment (jonbiemond) [reply bcitflex/app_functions/course_query.py:1]: Hey @MikeVarchola, yes you're right, if we can apply the filters to the select statement itself it would be more efficient. I did some research on how I could possible create a view using SQLAlchemy and Alembic and created issue #74

Comment (MikeVarchola) [reply bcitflex/app_functions/course_query.py:1]: @jonbiemond I can make simplified version of `.paginate()` on select()

it would look like this:
`def get_paginated_results(select, page, page_size):
    # Calculate offset
    offset = (page - 1) * page_size
    # Apply limit and offset
    paginated_select = select.limit(page_size).offset(offset)
    return paginated_select`

`page = 1
page_size = 25
query = select([courses])
paginated_select = get_paginated_results(select, page, page_size)`

but it would not work as good as original query.paginate()

Comment (jonbiemond) [reply bcitflex/app_functions/course_query.py:1]: Hey Mike, looking at the docs I see Flask-SQLAlchemy has a paginate function that takes a select statement as an argument. You can call it using `DBSession.paginate()`. https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/pagination/

I'll still need to implement SQL views as per #71 so we can introduce database level pagination without introducing some regression.

Comment (MikeVarchola) [reply bcitflex/app_functions/course_query.py:1]: @jonbiemond hey, I refactored the code, its using `select `instead of `query `now.

**Review by MikeVarchola — COMMENTED** [review]
@jonbiemond  pagination implemented for new solution, check it out please

**Comment by jonbiemond** [comment]
I still have to review, can you please change your PR to target the `dev` branch? I'll see if I can setup the tests to run without my approval.

**Comment by MikeVarchola** [comment]
> I still have to review, can you please change your PR to target the `dev` branch? I'll see if I can setup the tests to run without my approval.

DONE

**Review by Dominik26-1 — APPROVED** [review]

File: bcitflex/course.py
Comment (jonbiemond) [inline bcitflex/course.py:20]: Somehow after the change SQLAlchemy isn't able to join `Meeting` anymore. Could you please pass the association table to the call? 
```py
filters.where(Meeting.campus == request.args.get("campus"), links=[Offering])
```

File: tests/app/test_course.py
Comment (jonbiemond) [inline tests/app/test_course.py:1]: With the change to `GET`, can you please update the method used in the "filter_courses" tests?

**Review by jonbiemond — CHANGES_REQUESTED** [review]
Looking awesome @MikeVarchola! A couple of the database dependent tests failed when I ran them locally. If you like you can run all tests locally by installing postgres on your machine, there are [resources in the README.md](https://github.com/jonbiemond/BCIT-Flex#prerequisites).

File: bcitflex/course.py
Comment (MikeVarchola) [reply bcitflex/course.py:20]: @jonbiemond 

sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateAlias) table name "offering" specified more than once

[SQL: SELECT DISTINCT course.course_id, course.subject_id, course.code, course.name, course.prerequisites, course.credits, course.url, course.created_at, course.updated_at, course.deleted_at
FROM course JOIN offering ON course.course_id = offering.course_id AND offering.deleted_at IS NULL JOIN offering ON course.course_id = offering.course_id AND offering.deleted_at IS NULL
WHERE (offering.status NOT IN (%(status_1_1)s, %(status_1_2)s, %(status_1_3)s, %(status_1_4)s, %(status_1_5)s)) AND course.deleted_at IS NULL
 LIMIT %(param_1)s OFFSET %(param_2)s]
[parameters: {'param_1': 25, 'param_2': 25, 'status_1_1': 'Full', 'status_1_2': 'In Progress', 'status_1_3': 'Cancelled', 'status_1_4': 'Closed', 'status_1_5': 'Waitlist'}]


did you test out your code before merging ?

File: tests/app/test_course.py
Comment (MikeVarchola) [reply tests/app/test_course.py:1]: Done

File: bcitflex/course.py
Comment (MikeVarchola) [reply bcitflex/course.py:20]: @jonbiemond  you code does not work with 2+ conditions

Comment (jonbiemond) [reply bcitflex/course.py:20]: @MikeVarchola I see the problem, sorry about that. I'll add a test for that condition and put out a fix by EOD.

Comment (MikeVarchola) [reply bcitflex/course.py:20]: @jonbiemond  try filter for campus + filter for subject at same time

Comment (jonbiemond) [reply bcitflex/course.py:20]: @MikeVarchola I pushed a fix to dev with https://github.com/jonbiemond/BCIT-Flex/pull/76/commits/831fcddb8a4a7791e663971cf0959ed3cc9679cf .

**Review by jonbiemond — APPROVED** [review]
Fantastic, thanks so much @MikeVarchola! I'll rebase your changes onto my fix and resolve any conflicts locally.

**Pull request merged**
